### ISSUE TITLE
ofXml error on comment

### ISSUE THREAD

**Issue by turowskipaul** [openframeworks/openFrameworks#3760]
In the ofXml class, when I try to read through an XML file sequentially using the setToSibling() method, I get EXC_BAD_ACCESS from getValue() when a commented line, like
`<!--	<pRangeMin>36</pRangeMin>-->`
is encountered.

**Comment by bilderbuchi** [comment]
@joshuajnoble , any idea?

**Comment by tpltnt** [comment]
@turowskipaul I failed to replicate the bug as of c214832ee9a7e1812a249f0f731d620fc8e53000. I have tested different scenarios. Can you post example data and code which triggers the bug?

**Comment by turowskipaul** [comment]
Thanks for testing.  Here is some code:

testing.xml

    <testing>
        <one>1</one>
        <two>2</two>
        
        <four>4</four>
        <five>5</five>
    </testing>

ofApp.h

    #pragma once

    #include "ofMain.h"

    class ofApp : public ofBaseApp{
        ofXml myXml;
    public:
        void setup();
    };

ofApp.cpp

    #include "ofApp.h"
    void ofApp::setup(){
        if (myXml.load("testing.xml") ) {
            int numFields = myXml.getNumChildren();
        
            vector<float> xmlStorage(numFields);
        
            ofLogNotice() << "xmlStorage size: " << xmlStorage.size();
        
            myXml.setToChild(0);
            for (int x=0; x<numFields; x++) {
                xmlStorage[x] = myXml.getFloatValue();
                ofLogNotice() << x << " was set to: " << xmlStorage[x];

                myXml.setToSibling();
            }
        }
    }

Console output

    [notice ] xmlStorage size: 4
    [notice ] 0 was set to: 1
    [notice ] 1 was set to: 2
    (lldb) 

stops on string ofXml::getValue() const, with error:

    Thread 1: EXC_BAD_ACCESS (code=2, address 0x0)

The crash can be avoided with a simple test, like so:

ofApp.h

    #pragma once

    #include "ofMain.h"

    class ofApp : public ofBaseApp{
        ofXml myXml;
        static const string xmlNames[];
    
    public:
        void setup();
    };

ofApp.cpp

    #include "ofApp.h"

    const string ofApp::xmlNames[] = {
        "one",
        "two",
        "three",
        "four",
        "five"
    };

    //--------------------------------------------------------------
    void ofApp::setup(){
        if (myXml.load("testing.xml") ) {
            int numFields = myXml.getNumChildren();
        
            vector<float> xmlStorage(numFields);
        
            ofLogNotice() << "xmlStorage size: " << xmlStorage.size();
        
            myXml.setToChild(0);
            for (int x=0; x<numFields; x++) {
                string mName = myXml.getName();
                if (mName == xmlNames[x]) {
                    xmlStorage[x] = myXml.getFloatValue();
                    ofLogNotice("ofApp") << xmlNames[x] << " was set to: " << xmlStorage[x];
                }
                else {
                    xmlStorage[x] = 0;
                    ofLogError("ofApp") << "No value for " << xmlNames[x] << " in XML; setting to default: " << xmlStorage[x];
                }
            
                myXml.setToSibling();
            }
        }
    }

Console output

    [notice ] xmlStorage size: 4
    [notice ] ofApp: one was set to: 1
    [notice ] ofApp: two was set to: 2
    [ error ] ofApp: No value for three in XML; setting to default: 0
    [notice ] ofApp: four was set to: 4

However, it's still not storing the correct number of values because setToSibling() is stopping on the commented line.  In short, getNumChildren() and setToSibling() seem to be inconsistent.

**Comment by tpltnt** [comment]
@turowskipaul thank you for the data. I can replicate the bug for a7a6e5b3aa71c8aa06f39d8495d274cde026f8a0.

**Comment by tpltnt** [comment]
The data provided triggers multiple bugs . I fixed the NULL pointer dereference in getValue(). I am currently looking into the `setToSibling()` and `getNumChildren()` inconsistency.

**Comment by joshuajnoble** [comment]
Ah, I was going to look at this a little later today but it seems like you've got it handled so maybe I'll leave it to you?

**Comment by tpltnt** [comment]
I think i fixed this issue. I do have a personal test suite and ran the code posted which yields the following output:
```
[notice ] xmlStorage size: 4
[notice ] 0 was set to: 1
[notice ] 1 was set to: 2
[notice ] 2 was set to: 4
[notice ] 3 was set to: 5
```
@turowskipaul Can you check out [my branch](https://github.com/tpltnt/openFrameworks/tree/bugfix-issue3760) and check if your (other) code works now? (I want your ok as a basic quality check before issuing a pull request.)

**Comment by turowskipaul** [comment]
Worked for me. Cheers!

**Comment by tpltnt** [comment]
@turowskipaul please reopen since the issue still exits. My stuff has not been merged yet.

**Comment by turowskipaul** [comment]
sorry!

**Comment by tpltnt** [comment]
@turowskipaul no problem

**Comment by bilderbuchi** [comment]
closed by #4151
